design/export-functionality.md

```diff
@@ -0,0 +1,93 @@
+--conf spark.hadoop.lakefs.api.access_key=<LAKEFS_ACCESS_KEY_ID> \
+--conf spark.hadoop.lakefs.api.secret_key=<LAKEFS_SECRET_ACCESS_KEY> \
+--packages io.lakefs:lakefs-spark-client-301_2.12:0.1.0 \
+--class io.treeverse.clients.Main export-app example-repo s3://example-bucket/prefix \
+--branch=example-branch 
+``` 
+
+Pros:
+- Utilizes the existing spark client, and gives the same functionality. 
+ 
+Cons:
+- Require creating and maintaining the standalone container.
+
+### Using the java client
+Reimplement the spark client export functionality by creating a new Exporter class that doesn't use spark.
+This class can be part of the java cliet, or a new "Export" client that uses the java client.
+
+The usage will be with java code (in a similar way of using custom code with the spark client).
+
+In order to use Java code, users will need to download it if necessary, learn it, and have it in their toolset.
+
+Usage example:
+
+```java
+class Export{
+    public static void main(String[]args){
+        Exporter exporter = new Exporter(apiClient, repoName, rootLocation);
+        exporter.exportAllFromBranch(branchName);
+        exporter.exportFrom(branchName, commitRef);
+    }
+}
+```
+
+
+Pros:
+- An intermediate solution that is compatible with the issue.
+
+Cons:
+- Require checking how spark is being used in the current implementation (in order to rewrite its functions). Not necessarily possible in a naive way.
+- Require developing and maintaining a second export functionality.
+
+### Using Rclone
+[Rclone](https://rclone.org/) is a command line program to sync files and directories between cloud providers.
+Users can gain the export functionality by using Rclone's [copy](https://rclone.org/commands/rclone_copy/) and [sync](https://rclone.org/commands/rclone_sync/) commands.
+
+The copy command can be used to copy files from lakeFS to a designated object store location.
+The sync command can be used to export only the diff between a specific branch and a commit reference, since sync makes the source and the dest identical (modifying destination only).
+
+In order to use Rclone, users will need to download it if necessary, learn it, and have it in their toolset.
+
+Usage example:
+
+```shell
+rclone copy source:lakefs:example-repo/main/ dest:s3://example-bucket/prefix
+
+rclone sync source:lakefs:example-repo/main/ dest:s3://example-bucket/prefix
+```
+
+Pros:
+- Rclone gives the functionality of exporting data from lakeFS to a designated object store location. Require only few additional work of documentation for the usage of Rclone as an export functionality.
+- Simple to use. 
+- Doesn't require maintaining a new feature.
+
+Cons:
+- Doesn't support success/failure indications on the exported objects.
+- Require a designated object store location that doesn't contain any other data but the data associated to the lakeFS branch. That is because Rclone's sync command will delete all files that don't exist in the branch.   
```

Comment:
1. AFAIK deleting missing files is desired behaviour of export.
1. If you truly want only to copy,  won't `rclone copy` do the job?

Reply:
I meant that if you have files in the destination location that aren't related to the files in lakeFS - the sync command will delete those files. 
That is the reason users should allocate a designated object store location to export the files from lakeFS.

Reply:
As discussed f2f, this is either a bug (I kept files in the same "subtree" used for export, but didn't mean to delete those files!)  or a feature (after I export, I want the destination s3 prefix to look exactly like lakeFS does). You're not going to please everyone here.

FTR, what our Spark-based exporter does is a third thing, that relies on separately managing the "last exported revision". I f believe that only makes sense for large-scale installations - and the sync option is close enough if we document that the entire subtree is intended only for export.